### DISCUSSION TITLE
Databricks pyspark interoperability

### DISCUSSION THREAD

**Discussion by geoHeil** [dagster-io/dagster#18383] [Q&A] [answered]
- How can I make a dagster step launcher (or launched external job) in databricks compatible with pyspark for a fast local development experience?
- How can I override the step launcher in the local deployment to use the lodal pyspark instance instead of databricks?
- how can a monorrepo with an applicaiton project (main dagster pipeline) and a library project (reusable building blocks) easily be published over to databricks (without publishing the library to an artifact store) ? 
  - how should the `"local_pipeline_package_path": str(Path(__file__).parent.parent.parent),` be defined
  - especially in such a case I think this gets tricky
- are there some examples available (dagster graphs) how to update the dagster python script / move it to databricks? So far I have not seen an E2E (dagster automated) example.

it turns out for some of the points above some issues already exist:

- https://github.com/dagster-io/dagster/issues/18099

**Comment by zyd14** [comment]
> How can I override the step launcher in the local deployment to use the lodal pyspark instance instead of databricks?

On my team, if we have a DAG that we want to be able to execute locally as well as on Databricks, we'll maintain two jobs that are made from the same graph. In the locally-executed graph we'll replace the step launcher resource with `ResourceDefinition.none_resource()`
```python
from dagster import op, graph, ResourceDefinition
from dagster_databricks import databricks_pyspark_step_launcher

@op
def local_step():
  ...

@op(required_resource_keys={"step_launcher"})
def potentially_databricks_step(context, upstream):
  ...

@graph
def example():
  potentially_databricks_step(local_step())

production_job = example.to_job(resource_defs={"step_launcher": databricks_pyspark_step_launcher})
local_job = example.to_job(resource_defs={"step_launcher": ResourceDefinition.none_resource()})
```

> how can a monorrepo with an applicaiton project (main dagster pipeline) and a library project (reusable building blocks) easily be published over to databricks (without publishing the library to an artifact store) ? 

Regarding moving dependencies to the Databricks environment, Docker environments are the cleanest way to go. If you can publish your library / environment to ECR or Dockerhub and then point to it from your step launcher config it makes it much easier to manage. However, to my knowledge the Ganglia UI doesn't work with Docker environments, so you'll need another monitoring source for memory / cpu stats like Datadog (which is able to be implemented through an init script on the cluster). If you can't publish your dependencies as a docker image, you can also consider building a wheel from the library and pushing it to S3 or to your Databricks workspace as part of your CI, and using the `libraries` configuration (under the `run_config` option) of the step launcher to point to the wheel in S3 or in your Databricks workspace.

> are there some examples available (dagster graphs) how to update the dagster python script / move it to databricks? So far I have not seen an E2E (dagster automated) example.

The step launcher should already move the Dagster step that needs to be executed to the Databricks environment. I may be missing something about this question though.

**Reply by geoHeil** [reply]
What mechanism do you use to switch local and production job execution based on the environment?
Or do you simply render both jobs in dagit and the user must select the right one?

**Reply by geoHeil** [reply]
But your proposed example should also work for `@assets` right? https://docs.dagster.io/concepts/assets/graph-backed-assets with `@graph_asset`?

**Reply by geoHeil** [reply]
Is there a reason why you opted against such an approach:

```
if os.getenv("COMPUTE")=="LOCAL":
    resources = {"none_resource"}
else:
    resources = {"step_launcher"}

@asset(
    required_resource_keys=resources
)
```

**Reply by zyd14** [reply]
The main reason we render both on our team is because we do a lot of testing on Databricks from local Dagster environments, as we don't have great test datasets that can be run against locally. The approach of switching based on environments should work too. The approach works with `@assets` and `@graph_assets` as well.

**Reply by geoHeil** [reply]
Makes sense. But Wouldn`t you maybe want to separate testing / dev assets in branch deploys - perhaps even via a local dagster cloud runner for faster changes?

**Reply by geoHeil** [reply]
i just found this example: https://github.com/dagster-io/hooli-data-eng-pipelines/blob/master/hooli_data_eng/assets/forecasting/__init__.py#L148-L162

How is context.resources.pyspark.spark_session in PROD case use steplauncher? I do not observe any steplauncher interaciton here it looks like pyspark is used always and only.

How can this fallback mode work so seamless?

If this works so neat - why would you need the factory or two-graph approach (in case of the steplauncher approach)?

Observing that this works - I wonder how. And I am also seeing some issues around special spark options or loading of JARs if spark is launched in that way. At least for me it looks like the normal (initialized) pysparksession is not used and somehow magically a 2nd one is created. When additional flags like delta classes or jars would be present - these are missing now and result in class not found exceptions.

```

"pyspark": configured_pyspark,
    "step_launcher": ResourceDefinition.none_resource(),
    "parquet_io_manager": LocalPartitionedParquetIOManager(
        pyspark=configured_pyspark
    ),

```

I.e. an asset which is reading with the IO manager is instanciated without missing classes. The 

```
@op
def full_moon_dates_op():
    spark = SparkSession.builder.appName("FullMoonDates").getOrCreate()

    try:
        r = requests.get(FULL_MOON_DATES_URL, timeout=10)
        with open(CSV_FILE_PATH, "wb") as f:
            f.write(r.content)
    except requests.RequestException as e:
        print(f"Error while downloading the file: {e}")
        return

    full_moon_dates_df = spark.read.option("header", "true").csv(
        "seed_full_moon_dates.csv"
    )
    full_moon_dates_df = full_moon_dates_df.withColumn(
        "full_moon_date", col("full_moon_date").cast("date")
    )
    full_moon_dates_df.write.format("delta").mode("overwrite").saveAsTable(
        "hive_metastore.<DB_SCHEMA>.raw_full_moon_dates"
    )
@asset(required_resource_keys={"step_launcher", "pyspark"},io_manager_key="parquet_io_manager",)
def full_moon_dates_delta_table():
    full_moon_dates_op()
```

variant here fails with class not found

**Reply by zyd14** [reply]
I'm having a little trouble following your question here. Are you specifically asking how this executes using a local spark session when executed locally? And how it then switches to executing in Databricks when in dev or prod? If so, I think the key lies in the `definitions.py` file, where the [resources for the code location are defined](https://github.com/dagster-io/hooli-data-eng-pipelines/blob/master/hooli_data_eng/definitions.py#L65). It looks like they're switching whether the real step launcher or a none_resources is being used depending on the environment, similar to what you suggested [here](https://github.com/dagster-io/dagster/discussions/18383#discussioncomment-7714804).

I'm not sure why the java classes would be missing from the spark session in your example. One thing that I would do to try to eliminate possible issues would be to access the spark session through the resources passed to the op (like `context.resources.pyspark.spark_session`), instead of through `SparkSession.builder.appName("FullMoonDates").getOrCreate()`. It's unclear what you're using as your base `configured_pyspark` resource - if you're using the lazy implementation of the spark resource, then the spark session won't actually get instantiated until you access it through the resources, so `SparkSession.builder.appName("FullMoonDates").getOrCreate()` would be the first time the spark session is instantiated, without any configuration (because none is provided).

**Reply by geoHeil** [reply]
Indeed yes exactly. I want to be able to switch smoothly betwenn local, databricks-interactive and databricks-job clusters.

Your point regarding the session was right - I overlooked it. Thanks.

However:

- I still do not understand how the databricks (only) mode would work. There, I would presumably pass pyspark resource = None - and trying to access the pyspark resource would then fail
- When currently trying for the local mode the code fails with:

```
dagster._core.errors.DagsterInvalidInvocationError: Decorated function 'full_moon_dates_op' has context argument, but no context was provided when invoking.

```

I have modified the functions to pass and accept the contect:

```
@op(required_resource_keys={"step_launcher", "pyspark"})
def full_moon_dates_op(context):

@asset(
    required_resource_keys={"step_launcher", "pyspark"},
    io_manager_key="parquet_io_manager",
)
def full_moon_dates_delta_table(context):
    full_moon_dates_op(context)
```

Why is the context not passed in? As I am in fact explicitly trying to pass it

**Reply by zyd14** [reply, marked as answer]
This is just a Dagster thing - you shouldn't really be calling ops from assets, an asset is already an op under the hood. It will work okay in some cases, as long as you don't have arguments you want to pass to the op from the asset (particularly the context), but will bite you like this. If you want to share functionality between an asset and an op, then factor out a vanilla python function to call:
```python
@op(required_resource_keys={"step_launcher", "pyspark"})
def full_moon_dates_op(context):
  full_moon_dates_func(context)

def full_moon_dates_func(context):
  ...

@asset(
    required_resource_keys={"step_launcher", "pyspark"},
    io_manager_key="parquet_io_manager",
)
def full_moon_dates_delta_table(context):
    full_moon_dates_func(context)
```

I'm not sure why you would pass in pyspark resource = None when you want to execute in Databricks. All the pyspark resource does is abstract a SparkSession for you and inject it into an op or asset, so you should access the spark session through the pyspark resource regardless of whether you're executing locally or on Databricks. Locally it will handle configuring the spark session for you, on Databricks it will give you the spark session that Databricks instantiates for you (I think with any additional configuration you provide to the resource as well, as long as it's config that can be set after the spark session has been created - things like spark.jars.packages can only be set before the spark session is created the first time.) The formula for local vs. Databricks jobs is like so:
```python
local_job = job_graph.to_job(resource_defs={"step_launcher": databricks_pyspark_step_launcher, "pyspark": pyspark_resource})
databricks_job = job_graph.to_job(resource_defs={"step_launcher": ResourceDefinition.none_resource(), "pyspark": pyspark_resource})
```
if you want to switch between those using environment variables instead of having them both defined side-by-side, that's just an implementation detail, but the resource defs stay the same.

**Comment by geoHeil** [comment]
Has anyone explored untangling the step and job launcher job launch topic so it can be reconfigured to use i.e. local, databricks steplauncher a nd databricks job mode?

**Comment by smackesey** [comment]
I think @zyd14's responses are good if you are wedded to using `StepLauncher`, but have you considered trying out the dagster-pipes Databricks integration? https://docs.dagster.io/guides/dagster-pipes/databricks

You could potentially toggle local vs Databricks based execution by using some logic in your assets that changes the arguments to `open_pipes_session` depending on a config file, or maybe swaps in `PipesSubprocessClient` vs `PipesDatabricksClient`.

>are there some examples available (dagster graphs) how to update the dagster python script / move it to databricks? So far I have not seen an E2E (dagster automated) example.

This is a bit hacky but the tests we use for the pipes databricks integration upload both `dagster-pipes` and the test script on every run: https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-databricks/dagster_databricks_tests/test_pipes.py

**Reply by geoHeil** [reply]
I would love to get both modes (steplauncher for interactive DBU) but also pipes for single asset single cluster workflow DBU working. I am looking at both.

The change between PipesSubprocessClient and PipesDatabricksClient is looking as the smootheest for me (most compatible in behavior) as overrideing i.e. no step launcher with pyspark for local seems to be a bit different.

**Reply by geoHeil** [reply]
https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-databricks/dagster_databricks/_test_utils.py#L32 `upload_dagster_pipes_whl` thanks for the example. 

@smackesey https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-databricks/dagster_databricks/_test_utils.py#L43C14-L43C21 you are using `WorkspaceClient` plus `dbfs`. Does this imply that both need to be authorized separately? I.e. where do you configure the dbfs workspace/cluster settings? Shouldn`t it also be possible to perform this operation straight from the already authorized and configured `WorkspaceClient`?
